metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -256,101 +173,65 @@ ::call_outgoing_kmers(node_index node, const OutgoingEdgeCallback &callback) con
-        append_next_rc_nodes(node, children);
-
-    for (size_t c = 0; c < children.size(); ++c) {
-        if (children[c] != npos) {
-            callback(children[c], alphabet[c]);
-            assert(traverse(node, alphabet[c]) == children[c]);
-        }
-    }
-}
-
-void CanonicalDBG::append_prev_rc_nodes(node_index node,
-                                        SmallVector<node_index> &parents) const {
-    /**
-     * find parents of node by searching for children of its reverse complement
-     * e.g., node = AGCCAT. Find TAGCCA and AAGCCA by looking for TGGCTA and TGGCTT.
-     *  TAGCCA                    TGGCTA
-     *        \                  /
-     *         AGCCAT  ->  ATGGCT
-     *        /                  \
-     *  AAGCCA                    TGGCTT
-     */
-
-    // for each n, check for TGGCTn. If found, define and store the index for
-    // rc(n)AGCCA as index(TGGCTn) + offset_
-
-    if (const DBGSuccinct *dbg_succ = get_dbg_succ(*graph_)) {
-        const boss::BOSS &boss = dbg_succ->get_boss();
-
-        //   AGAGGATCTCGTATGCCGTCTTCTGCTTGAG
-        //-> AGAGGATCTCGTATGCCGTCTTCTGCTTGA
-        //-> TCAAGCAGAAGACGGCATACGAGATCCTCT
-        std::string rev_seq = graph_->get_node_sequence(node).substr(0, boss.get_k());
-        if (rev_seq[0] == boss::BOSS::kSentinel)
-            return;
-
-        ::reverse_complement(rev_seq.begin(), rev_seq.end());
-        auto encoded = boss.encode(rev_seq);
-        auto [rc_edge, edge_2, end] = boss.index_range(encoded.begin(), encoded.end());
-
-        if (end != encoded.end())
-            return;
-
-        assert(rc_edge == edge_2);
-
-        boss.call_outgoing(rc_edge, [&](boss::BOSS::edge_index adjacent_edge) {
-            assert(dbg_succ);
-            node_index prev = dbg_succ->boss_to_kmer_index(adjacent_edge);
-            if (prev == npos)
-                return;
+    if (max_num_edges_left) {
+        /**
+         *
+         * find children of node by searching for parents of its reverse complement
+         * e.g., node = ATGGCT. Find TGGCTA and TGGCTT by looking for  TAGCCA and AAGCCA
+         *         TGGCTA      TAGCCA
+         *        /                  \
+         *  ATGGCT         ->         AGCCAT
+         *        \                  /
+         *         TGGCTT      AAGCCA
+         */
+
+        // for each n, check for nAGCCA. If found, define and store the index for
+        // TGGCTrc(n) as index(nAGCCA) + offset_
+        assert(get_extension<NodeRC<>>());
+
+        const DBGSuccinct *dbg_succ = get_dbg_succ(*graph_);
+        const boss::BOSS *boss = dbg_succ ? &dbg_succ->get_boss() : nullptr;
+        const auto cache = get_extension<NodeFirstCache>();
 
-            boss::BOSS::TAlphabet c = boss.get_W(adjacent_edge) % boss.alph_size;
+        get_extension<NodeRC<>>()->call_incoming_nodes_from_rc(node, [&](node_index next) {
+            char c;
+            if (cache) {
+                c = cache->get_first_char(next);
```

Comment:
So is it helpful to have both cache and NodeRC? Are they doing different things?

Reply:
Sure. In CanonicalDBG, having the cache is still useful for traversing backwards in the graph, whereas this index is useful for quickly determining if a k-mer orientation switch can be done.